### DISCUSSION TITLE
Welcome to oneDPL Discussions!

### DISCUSSION THREAD

**Discussion by jiejanezhang** [uxlfoundation/oneDPL#126] [Q&A]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by michaeldmello** [comment]
Hello,

Can the oneDPL APIs or STL APIs be called from within a SYCL command group and not specifically from a parallel_for() launch?

thanks 

-michael

**Reply by jinge90** [reply]
Hi, @michaeldmello 
If we call a function in SYCL command group but not from parallel_for or single_task, the function called will be executed in host. If we call a function in parallel_for/single_task, the function will be executed in device side. In host, we can call any STL functions so we can call any STL functions within SYCL command group only if those functions are not called from a parallel_for/single_task launch.
For example:
deviceQueue.submit([&](cl::sycl::handler &cgh) {
        auto res_access = buffer1.get_access<sycl_write>(cgh);
        std::cout << "XXXXXXXX" << std::endl;  // iostream will do the print in host side.
        cgh.single_task<class KernelTest>([=]() {
          res_access[0] = std::sin(0.5f); // this std::sin will be executed in device.
        });
    });

The code above will print "XXXXXXXX" in screen, the print is done in host side.
Hi, @capatober
For PSTL algorithm using device policy, can we call those parallel algorithm in SYCL command group? 
Thanks very much.

**Reply by andreyfe1** [reply]
Actually, no. We call `submit()` in our backend for algorithms. I believe calling SYCL command group within another one is not allowed.

**Comment by abagusetty** [comment]
Hi, 

Are calls to PSTL using device policy async with respect to host? Or Does one needs to use some sort of `queue.wait()` after a given PSTL API call to enable completion ?

Thanks
-Abhishek

**Reply by andreyfe1** [reply]
Hi @abagusetty,
Algorithms are blocking by default. You don't need to call `wait()` explicitly. Are you looking for async abilities with respect to host?

**Reply by abagusetty** [reply]
Hey @capatober 
Thanks so much, in other words all the events/queue associated with the algorithms are blocked with `wait()`. 

Yes, I am exploring to see on how one can invoke these PSTL algos in both sync and async manner with respect to host. Do you happen to have any pointers on that.

Cheers!

**Reply by andreyfe1** [reply]
Sure.
1. Sync execution. It's by default as it's mentioned above.
2. Quasi async execution. Use `ONEDPL_ALLOW_DEFERRED_WAITING` macro to call `queue.wait()` in that part of code when you need it. It's quasi async because it doesn't work for any algorithm. For example, if you call `std::reduce`, it returns a value. We need to get this value from the device. For that we need to wait till a device completes the execution. So, deferred waiting is not applicable here. [Link to the documentation](https://docs.oneapi.com/versions/latest/onedpl/pstl/macros.html).
3. Fully async execution. This feature is experimental, just a small subset of algorithms is supported for now. `std::reduce` call looks like 
```
auto ret_val = dpl::experimental::reduce_async(dpl::execution::dpcpp_default, dpl::begin(a),dpl::end(a));
...
auto deferred_value = ret_val.get();
```
[Link to the documentation](https://github.com/oneapi-src/oneDPL/blob/main/documentation/library_guide/extension_api.rst#async-api).

Does the information meet your expectations?

**Reply by abagusetty** [reply]
@capatober Thank you so much for the documentation. This is really helpful with planning.

**Comment by abagusetty** [comment]
Hi !

Can some one provide pointers/snippets on how to accomplish ops like `scatter_if` and `gather_if` using a combination of dpl and stl algorithms/iterators efficiently for device-side execution.

**Comment by christosa-rfpro** [comment]
Hi,

It's not so clear which (if any) variants of `std::inclusive_scan` and `std::transform_inclusive_scan` have parallel implementations when using the OpenMP backend (i.e. when `ONEDPL_USE_OPENMP_BACKEND` is set to `1`).

Can you please clarify?

Thanks

**Reply by MikeDvorskiy** [reply]
Hello @chrisakriv-rfpro,

Unfortunately, the OpenMP backend has still not parallel implementation for `std::inclusive_scan`, `std::exclusive_scan`, `std::transform_inclusive_scan`, `std::transform_exclusive_scan`. You get a serial execution for the mentioned algorithms.
